python/ray/tests/kuberay/utils.py

```diff
@@ -0,0 +1,151 @@
+    """Waits for the pod to enter running Running status.phase."""
+    for i in range(tries):
+        pod_status = (
+            subprocess.check_output(
+                [
+                    "kubectl",
+                    "-n",
+                    namespace,
+                    "get",
+                    "pod",
+                    pod,
+                    "-o",
+                    "custom-columns=POD:status.phase",
+                    "--no-headers",
+                ]
+            )
+            .decode()
+            .strip()
+        )
+        # "not found" is part of the kubectl output if the pod's not there.
+        if "not found" in pod_status:
+            raise Exception(f"Pod {pod} not found.")
+        elif pod_status == "Running":
+            logger.info(f"Confirmed pod {pod} is Running.")
+            return
+        elif i < tries - 1:
+            logger.info(
+                f"Pod {pod} has status {pod_status}. Waiting for the pod to enter "
+                "Running status."
+            )
+            time.sleep(backoff_s)
+        else:
+            raise Exception(f"Timed out waiting for pod {pod} to enter Running status.")
+
+
+def wait_for_ray_health(ray_pod: str, namespace: str, tries=60, backoff_s=5) -> None:
+    """Waits for the Ray pod to pass `ray health-check`.
+    (Ensures Ray has completely started in the pod.)
+    """
+    for i in range(tries):
+        try:
+            kubectl_exec(["ray", "health-check"], ray_pod, namespace)
+            logger.info(f"ray health check passes for pod {ray_pod}")
+            return
+        except subprocess.CalledProcessError as e:
+            logger.info(f"Failed ray health check for pod {ray_pod}.")
+            if i < tries - 1:
+                logger.info("Trying again.")
+                time.sleep(backoff_s)
+            else:
+                logger.info("Giving up.")
+                raise e from None
+
+
+def get_pod(pod_name_filter: str, namespace: str) -> str:
+    """Gets pods in the `namespace`.
+
+    Returns the first pod that has `pod_name_filter` as a
+    substring of its name. Raises an assertion error if there are no matches.
+    """
+    get_pods_output = (
+        subprocess.check_output(
+            [
+                "kubectl",
+                "-n",
+                namespace,
+                "get",
+                "pods",
+                "-o",
+                "custom-columns=POD:metadata.name",
+                "--no-headers",
+            ]
+        )
+        .decode()
+        .split()
+    )
+    matches = [item for item in get_pods_output if pod_name_filter in item]
+    assert matches, f"No match for `{pod_name_filter}` in namespace `{namespace}`."
+    return matches[0]
+
+
+def kubectl_exec(command: List[str], pod: str, namespace: str) -> None:
+    """kubectl exec the `command` in the given `pod` in the given `namespace`"""
+    subprocess.check_call(["kubectl", "exec", "-it", pod, "--"] + command)
```

Comment:
This will work for most of the time. Sometimes we may have multiple containers and without specifying the container `-c ${container_name}`, it may go to unexpected ones. The good thing is the pattern is pretty fixed here.

Head will have `ray-head` and `autoscaler` and `ray-head` is the first container which will be chosen.  Worker has additional init-container which is not a problem.

`Defaulted container "ray-worker" out of: ray-worker, init-myservice (init)`

Reply:
That's a good point. It would be better to allow specifying a container here -- I will make that change in this PR.

Reply:
Added a container arg.